geoapps_utils/numerical.py

```diff
@@ -71,3 +73,103 @@ def traveling_salesman(locs: np.ndarray) -> np.ndarray:
         mask[current] = False
 
     return np.asarray(order)
+
+
+def filter_curves(curves, min_length, min_angle) -> list[list[list[float]]]:
+    """
+    Filter curves based on length and angle.
+
+    :param curves: List of curves.
+    :param min_length: Minimum number of points in a curve.
+    :param min_angle: Minimum angle between points in a curve, in radians.
+
+    :return: List of filtered curves.
+    """
+    filtered_curves = []
+    sub_curves = []
+    for curve in curves:
+        # Check that the curve is long enough
+        if len(curve) >= min_length:
+            # Check that the angles are not too sharp
+            for i in range(len(curve) - 2):
+                vec1 = [curve[i][0] - curve[i + 1][0], curve[i][1] - curve[i + 1][1]]
+                vec2 = [
+                    curve[i + 2][0] - curve[i + 1][0],
+                    curve[i + 2][1] - curve[i + 1][1],
+                ]
+
+                angle = 2 * np.pi - (
+                    np.arctan2(vec2[1], vec2[0]) - np.arctan2(vec1[1], vec1[0])
+                )
+
+                left_bound = min_angle
+                right_bound = 2 * np.pi - min_angle
+
+                if not left_bound <= angle <= right_bound:
+                    if i + 2 >= min_length:
+                        sub_curves.append(curve[: i + 2])
+                    if len(curve) - (i + 2) >= min_length:
+                        sub_curves.append(curve[i + 1 :])
+                    break
+
+                if i == (len(curve) - 3):
+                    filtered_curves.append(curve)
+
+    if len(sub_curves) > 0:
+        filtered_curves += filter_curves(sub_curves, min_length, min_angle)
+    return filtered_curves
+
+
+def find_curves(  # pylint: disable=too-many-locals
+    points: Points, min_length: int, max_distance: float, min_angle: float
+) -> list[list[list[float]]]:
+    """
+    Find curves in a set of points.
+
+    :param points: Points object to find curves in.
+    :param min_length: Minimum number of points in a curve.
+    :param max_distance: Maximum distance between points in a curve.
+    :param min_angle: Minimum angle between points in a curve, in radians.
+
+    :return: List of curves.
+    """
+    point_vertices = points.vertices
+    point_vertices = np.delete(point_vertices, 2, axis=1)
+    line_ids = points.get_data("line_ids")[0].values
+    channels = points.get_data("channel_group")[0].values
```

Comment:
We can probably avoid worrying about the groups if we do this operation one per...